File: src/themes/black-atom-engineering.ts
import * as Theme from "../types/theme";

const meta: Theme.Meta = {
  key: "black-atom-engineering",
  label: "Black Atom - Default [ENG]",
  appearance: "dark",
  status: "release",
  collection: {
    key: "default",
    label: "Default"
  }
};

const primaries: Theme.Primaries = [
  // Lows
  "#040505",
  "#0e1110",
  "#181b19",
  "#1d201e",

  // Mids
  "#3e4441",
  "#565f59",
  "#606a65",
  "#79857f",

  // Highs
  "#9fafa7",
  "#abbbb3",
  "#c4d4cc",
  "#d0e0d8"
];

const palette: Theme.Palette = {
  black: primaries[2],
  gray: primaries[5],

  darkRed: "#9891d6",
  red: "#a6a1df",

  darkGreen: "#93c270",
  green: "#a0ce7d",

  darkYellow: primaries[9],
  yellow: "#9BE197",

  darkBlue: primaries[8],
  blue: primaries[10],

  darkMagenta: "#a498e9",
  magenta: "#b7aded",

  darkCyan: "#6ac779",
  cyan: "#88dc96",

  lightGray: primaries[8],
  white: primaries[11]
};

const ui: Theme.UI = {
  bg: {
    default: primaries[2],
    panel: primaries[1],
    float: primaries[1],
    active: primaries[3],
    disabled: primaries[5],
    hover: primaries[3],
    selection: primaries[4],
    search: primaries[4],
    contrast: primaries[11],
    negative: palette.red,
    warn: palette.yellow,
    info: palette.blue,
    hint: palette.darkYellow,
    positive: palette.green,
    add: palette.green,
    delete: palette.red,
    modify: palette.blue
  },
  fg: {
    default: primaries[11],
    subtle: primaries[9],
    disabled: primaries[7],
    accent: palette.yellow,
    contrast: primaries[2],
    negative: palette.red,
    warn: palette.yellow,
    info: palette.blue,
    hint: palette.darkYellow,
    positive: palette.green,
    add: palette.green,
    delete: palette.red,
    modify: palette.blue
  }
};

const syntax: Theme.Syntax = {
  variable: {
    default: palette.blue,
    builtin: palette.blue,
    member: palette.darkBlue,
    parameter: palette.darkYellow
  },
  property: {
    default: palette.darkBlue
  },
  string: {
    default: palette.green,
    doc: palette.green,
    regexp: palette.red,
    escape: palette.red
  },
  constant: {
    default: primaries[10],
    builtin: palette.darkRed
  },
  module: {
    default: palette.darkBlue
  },
  boolean: {
    default: palette.darkGreen
  },
  number: {
    default: palette.darkGreen
  },
  type: {
    default: palette.cyan,
    builtin: palette.darkCyan
  },
  attribute: {
    default: palette.darkYellow,
    builtin: palette.darkYellow
  },
  func: {
    default: palette.yellow,
    builtin: palette.yellow,
    method: palette.yellow
  },
  constructor: {
    default: palette.yellow
  },
  keyword: {
    default: palette.magenta,
    import: palette.red,
    export: palette.red
  },
  operator: {
    default: palette.magenta
  },
  punctuation: {
    default: palette.lightGray,
    delimiter: primaries[8],
    bracket: palette.lightGray,
    special: palette.lightGray
  },
  comment: {
    default: palette.gray,
    doc: palette.darkGreen,
    todo: palette.green,
    error: palette.red,
    warn: palette.yellow,
    info: palette.blue,
    hint: palette.darkYellow
  },
  markup: {
    heading: {
      h1: palette.yellow,
      h2: palette.yellow,
      h3: palette.yellow,
      h4: palette.lightGray,
      h5: palette.lightGray,
      h6: palette.lightGray
    },
    list: {
      default: palette.lightGray,
      checked: palette.green,
      unchecked: palette.lightGray
    },
    quote: palette.green,
    math: palette.darkGreen,
    link: palette.green,
    code: {
      fg: palette.gray,
      bg: palette.black
    }
  },
  tag: {
    default: palette.yellow,
    builtin: palette.darkYellow,
    attribute: primaries[8],
    delimiter: primaries[9]
  }
};

const theme: Theme.Definition = {
  meta,
  primaries,
  palette,
  ui,
  syntax
};

export default theme;

File: ft/lua/black-atom-engineering.lua
local M = {}

---@class BlackAtom.Theme.Meta
M.meta = {
    key = "black-atom-engineering",
    label = "Black Atom - Default [ENG]",
    appearance = "dark",
    status = "release",
    collection = {
        key = "default",
        label = "Default",
    },
}

---@class BlackAtom.Theme.Primaries
M.primaries = {
    "#040505",
    "#0e1110",
    "#181b19",
    "#1d201e",
    "#3e4441",
    "#565f59",
    "#606a65",
    "#79857f",
    "#9fafa7",
    "#abbbb3",
    "#c4d4cc",
    "#d0e0d8",
}

---@class BlackAtom.Theme.Palette
M.palette = {
    black = "#181b19",
    gray = "#565f59",
    dark_red = "#9891d6",
    red = "#a6a1df",
    dark_green = "#93c270",
    green = "#a0ce7d",
    dark_yellow = "#abbbb3",
    yellow = "#9BE197",
    dark_blue = "#9fafa7",
    blue = "#c4d4cc",
    dark_magenta = "#a498e9",
    magenta = "#b7aded",
    dark_cyan = "#6ac779",
    cyan = "#88dc96",
    light_gray = "#9fafa7",
    white = "#d0e0d8",
}

---@class BlackAtom.Theme.UI
M.ui = {
    bg = {
        default = "#181b19",
        panel = "#0e1110",
        float = "#0e1110",
        active = "#1d201e",
        disabled = "#565f59",
        hover = "#1d201e",
        selection = "#3e4441",
        search = "#3e4441",
        contrast = "#d0e0d8",
        negative = "#a6a1df",
        warn = "#9BE197",
        info = "#c4d4cc",
        hint = "#abbbb3",
        positive = "#a0ce7d",
        add = "#a0ce7d",
        delete = "#a6a1df",
        modify = "#c4d4cc",
    },
    fg = {
        default = "#d0e0d8",
        subtle = "#abbbb3",
        accent = "#9BE197",
        disabled = "#79857f",
        contrast = "#181b19",
        negative = "#a6a1df",
        warn = "#9BE197",
        info = "#c4d4cc",
        hint = "#abbbb3",
        positive = "#a0ce7d",
        add = "#a0ce7d",
        delete = "#a6a1df",
        modify = "#c4d4cc",
    },
}

---@class BlackAtom.Theme.Syntax
M.syntax = {
    variable = {
        default = "#c4d4cc",
        builtin = "#c4d4cc",
        parameter = "#abbbb3",
        member = "#9fafa7"
    },
    constant = {
        default = "#c4d4cc",
        builtin = "#9891d6",
    },
    module = {
        default = "#9fafa7",
    },
    string = {
        default = "#a0ce7d",
        doc = "",
        regexp = "#a6a1df",
        escape = "#a6a1df"
    },
    boolean = {
        default = "#93c270"
    },
    number = {
        default = "#93c270"
    },
    type = {
        default = "#88dc96",
        builtin = "#6ac779"
    },
    attribute = {
        default = "#abbbb3",
        builtin = "#abbbb3"
    },
    func = {
        default = "#9BE197",
        builtin = "#9BE197",
        method = "#9BE197"
    },
    constructor = {
        default = "#9BE197"
    },
    operator = {
        default = "#b7aded"
    },
    keyword = {
        default = "#b7aded"
    },
    punctuation = {
        default = "#9fafa7",
        delimiter = "#9fafa7",
        bracket = "#9fafa7",
        special = "#9fafa7"
    },
    comment = {
        default = "#565f59",
        doc = ""
    },
    markup = {
        heading = {
            h1 = "#9BE197",
            h2 = "#9BE197",
            h3 = "#9BE197",
            h4 = "#9fafa7",
            h5 = "#9fafa7",
            h6 = "#9fafa7",
        },
        strong = "#9BE197",
        italic = "#9BE197",
        strikethrough = "#9BE197",
        quote = "#a0ce7d",
        math = "#93c270",
        link = "#a0ce7d",
        code = {
            fg = "#565f59",
            bg = "#181b19",
        },
        list = {
            default = "#9fafa7",
            checked =  "#a0ce7d",
            unchecked = "#9fafa7"
        },
    },
    tag = {
        default = "#9BE197",
        builtin = "#abbbb3",
        attribute = "#9fafa7",
        delimiter = "#abbbb3"
    }
}

---@class BlackAtom.Theme.Colors
M.colors = {
    none = "none",
    debug = M.debug,
    primaries = M.primaries,
    palette = M.palette,
    semantics = M.semantics,
    syntax = M.syntax,
}

return M
